File: Codigo/Front/Car2/face_recognition.py
import os
import time
from flask import Flask, request, jsonify
from dotenv import load_dotenv
from azure.cognitiveservices.vision.face import FaceClient
from msrest.authentication import CognitiveServicesCredentials
import cv2

# Carregar variáveis de ambiente
load_dotenv()
face_endpoint = os.getenv('FACE_ENDPOINT')
face_api_key = os.getenv('FACE_APIKEY')

# Autenticação com a API do Azure
credentials = CognitiveServicesCredentials(face_api_key)
face_client = FaceClient(face_endpoint, credentials)

person_group_id = 'my-unique-person-group'

app = Flask(__name__)

# Criar um grupo de pessoas
@app.route('/create_person_group', methods=['POST'])
def create_person_group():
    try:
        face_client.person_group.create(person_group_id, name=person_group_id)
        return jsonify({"message": "Person group created successfully."}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Adicionar pessoas ao grupo
@app.route('/add_faces', methods=['POST'])
def add_faces_to_person_group():
    try:
        image_base_url = request.json['image_base_url']
        person_dictionary = request.json['person_dictionary']
        
        for name, images in person_dictionary.items():
            person = face_client.person_group_person.create(person_group_id, name=name)
            for image in images:
                url = f'{image_base_url}{image}'
                face_client.person_group_person.add_face_from_url(person_group_id, person.person_id, url)
                
        return jsonify({"message": "Faces added to person group successfully."}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Treinar o grupo de pessoas
@app.route('/train_person_group', methods=['POST'])
def train_person_group():
    try:
        face_client.person_group.train(person_group_id)
        while True:
            status = face_client.person_group.get_training_status(person_group_id)
            if status.status == 'succeeded':
                break
            elif status.status == 'failed':
                raise Exception('Training failed')
            time.sleep(1)
        return jsonify({"message": "Person group trained successfully."}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Identificar rostos em uma imagem
@app.route('/identify_faces', methods=['POST'])
def identify_faces():
    try:
        image_path = request.json['image_path']
        with open(image_path, 'rb') as image:
            detected_faces = face_client.face.detect_with_stream(image, detection_model='detection_03')
        face_ids = [face.face_id for face in detected_faces]

        if face_ids:
            results = face_client.face.identify(face_ids, person_group_id)
            identified_faces = []
            for result in results:
                if result.candidates:
                    person = face_client.person_group_person.get(person_group_id, result.candidates[0].person_id)
                    identified_faces.append({"person_name": person.name, "confidence": result.candidates[0].confidence})
                else:
                    identified_faces.append({"face_id": result.face_id, "message": "No person identified"})
            return jsonify({"identified_faces": identified_faces}), 200
        else:
            return jsonify({"message": "No faces detected."}), 400
    except Exception as e:
        return jsonify({"error": str(e)}), 500

if __name__ == '__main__':
    app.run(debug=True)

File: Codigo/Front/Car2/faceRecognition.js
require('dotenv').config();
const { CognitiveServicesCredentials } = require("@azure/ms-rest-js");
const Face = require("@azure/cognitiveservices-face");
const { v4: uuidv4 } = require("uuid");

const endpoint = process.env.FACE_ENDPOINT;
const apiKey = process.env.FACE_APIKEY;
const credentials = new CognitiveServicesCredentials(apiKey);
const client = new Face.FaceClient(credentials, endpoint);
const personGroupId = uuidv4();

async function createPersonGroup() {
  console.log(`Creating a person group with ID: ${personGroupId}`);
  await client.personGroup.create(personGroupId, { name: personGroupId, recognitionModel: "recognition_04" });
}

async function addFacesToPersonGroup() {
  const imageBaseUrl = "https://raw.githubusercontent.com/Azure-Samples/cognitive-services-sample-data-files/master/Face/images/";
  const personDictionary = {
    "Family1-Dad": ["Family1-Dad1.jpg", "Family1-Dad2.jpg"],
    "Family1-Mom": ["Family1-Mom1.jpg", "Family1-Mom2.jpg"],
    "Family1-Son": ["Family1-Son1.jpg", "Family1-Son2.jpg"],
    "Family1-Daughter": ["Family1-Daughter1.jpg", "Family1-Daughter2.jpg"],
    "Family2-Lady": ["Family2-Lady1.jpg", "Family2-Lady2.jpg"],
    "Family2-Man": ["Family2-Man1.jpg", "Family2-Man2.jpg"],
  };

  for (const [name, images] of Object.entries(personDictionary)) {
    console.log(`Creating person: ${name}`);
    const person = await client.personGroupPerson.create(personGroupId, { name });
    for (const image of images) {
      const url = `${imageBaseUrl}${image}`;
      console.log(`Adding face to ${name} from image ${image}`);
      await client.personGroupPerson.addFaceFromUrl(personGroupId, person.personId, { url });
    }
  }
}

async function trainPersonGroup() {
  console.log(`Training person group: ${personGroupId}`);
  await client.personGroup.train(personGroupId);
  await waitForTraining();
}

async function waitForTraining() {
  while (true) {
    const status = await client.personGroup.getTrainingStatus(personGroupId);
    console.log(`Training status: ${status.status}`);
    if (status.status === "succeeded") {
      break;
    } else if (status.status === "failed") {
      throw new Error("Training failed");
    }
    await new Promise((resolve) => setTimeout(resolve, 1000));
  }
}

async function identifyFaces() {
  const imageBaseUrl = "https://raw.githubusercontent.com/Azure-Samples/cognitive-services-sample-data-files/master/Face/images/";
  const image = `${imageBaseUrl}identification1.jpg`;
  const detectedFaces = await client.face.detectWithUrl(image, { detectionModel: "detection_03", recognitionModel: "recognition_04" });
  const faceIds = detectedFaces.map(face => face.faceId);

  if (faceIds.length > 0) {
    const results = await client.face.identify(faceIds, { personGroupId });
    for (const result of results) {
      if (result.candidates.length > 0) {
        const person = await client.personGroupPerson.get(personGroupId, result.candidates[0].personId);
        console.log(`Person identified: ${person.name}, Confidence: ${result.candidates[0].confidence}`);
      } else {
        console.log(`No person identified for face ID ${result.faceId}`);
      }
    }
  } else {
    console.log("No faces detected.");
  }
}

async function main() {
  try {
    await createPersonGroup();
    await addFacesToPersonGroup();
    await trainPersonGroup();
    await identifyFaces();
  } catch (error) {
    console.error(error);
  } finally {
    console.log(`Deleting person group: ${personGroupId}`);
    await client.personGroup.deleteMethod(personGroupId);
    console.log("Done.");
  }
}

main();
